plutus-tx/src/PlutusTx/IsData/TH.hs

```diff
@@ -57,6 +58,74 @@ mkConstrPartsMatchPattern conIx extractFieldNames =
 
 -- TODO: safe match for the whole thing? not needed atm
 
+-- | Generate a function that matches on a 'BuiltinData' value and decodes it as a product type.
+mkAsDataMatchingFunction
+  :: TH.Name
+  -- ^ The name of the type
+  -> [TH.Name]
+  -- ^ Type variables of the type
+  -> TH.Name
+  -- ^ The name of the constructor
+  -> [TH.Type]
+  -- ^ Types of the fields
+  -> [TH.Name]
+  -- ^ The names of the fields
+  -> TH.Q (TH.Dec, TH.Dec)
+mkAsDataMatchingFunction name typeVars consName fieldTypes fields = do
+  let numFields = length fields
+  funcName <- TH.newName $ "matchOn" <> TH.nameBase name
+  builtinData <- TH.newName "builtinData"
+  asConstrN <- TH.newName "asConstr"
+  constrArgsN <- TH.newName "constrArgs"
+  restNs <- traverse (\i -> TH.newName $ "rest" <> show i) [0 .. numFields - 2]
+  fieldNs <- traverse (\i -> TH.newName $ "field" <> show i) [0 .. numFields - 1]
+  let argPat = TH.ConP consName [] [TH.VarP builtinData]
+      restDecs =
+        flip fmap (zip restNs (constrArgsN : restNs)) $ \(resti, restj) ->
+          TH.ValD
+            (TH.VarP resti)
+            (TH.NormalB $ TH.AppE (TH.VarE 'BI.tail) (TH.VarE restj))
+            []
+      fieldDecs =
+          flip fmap (zip fieldNs (constrArgsN : restNs)) $ \(fieldi, restj) ->
+              TH.ValD
+                (TH.VarP fieldi)
+                (TH.NormalB
+                  $ TH.AppE
+                      (TH.VarE 'unsafeFromBuiltinData)
+                      (TH.AppE
+                        (TH.VarE 'BI.head)
+                        (TH.VarE restj)
+                      )
+                )
+                []
+      decs =
+        [ TH.ValD
+            (TH.VarP asConstrN)
+            (TH.NormalB $ TH.AppE (TH.VarE 'BI.unsafeDataAsConstr) (TH.VarE builtinData))
+            []
+        , TH.ValD
+            (TH.VarP constrArgsN)
+            (TH.NormalB $ TH.AppE (TH.VarE 'BI.snd) (TH.VarE asConstrN))
+            []
+        ]
+        <> fieldDecs
+        <> restDecs
```

Comment:
Another very strange thing: if I reverse the order here, the tests change and report worse performance! So the only difference is that the `rest...` variables are declared first, and I imagine the order of declarations should not affect performance or semantics. I didn't look into this too much, but it's something that we might want to concern ourselves with.

Reply:
That's very strange. We should investigate by shrinking the test case and finding a small reproducible example. Please make an issue.